### PR TITLE
Add watchOptions to tsconfig and allow supplying them on command line as well

### LINKED ISSUE #30208: 100% CPU when installing node_modules while a file is open
**TypeScript Version:**  typescript@3.4.0-dev.20190302

**Search Terms:** tsserver cpu ember reload

**Steps to reproduce**

This is super-verbose; the relevant steps are bold and should be easy to reproduce

- Tested in MacOS
- Clone https://github.com/phil294/tsserver_lag_demo2. (It includes just a few dependencies with which the issue can be reproduced)
- Open it up in vscode
- Open process explorer so you can see the CPU usage of `tsserver`
- Do *not* open `main.js` (if you already have, close the tab & reload)
- Run `rm -rf node_modules && yarn`
- Wait for it to finish
- Open `main.js`
- `tsserver` cpu goes up to 50% for ~1 second. *This is how it should be*
- **Keep `main.js` open**
- **Run `rm -rf node_modules && yarn`**
- **Wait for it to finish**

**Expected behavior:**

`tsserver` should not go mad

**Actual behavior:**

`tsserver` cpu goes up to 100% and stays like this for a long time. It varies, but I've had both 30 seconds and 15 minutes. It renders the entire intellisense and go-to mechanics useless. Reinstalling `node_modules` is one of the most common things to do in a JS/TS project. This bug makes developing Ember.js applications with vscode & typescript intellisense support effectively impossible. For example, in our team, we all had to go back to Sublime Text because of this. The CPU usage is probably related to some dependency in package.json, but I have not been able to narrow it down. Either way, it is a Typescript bug.

**Related Issues:**
There are multiple tsserver CPU related issues. [This](https://github.com/Microsoft/TypeScript/issues/25629) one was the most similar but is missing details and is also closed, so probably not related.

### PR DESCRIPTION
**Pull request by sheetalkamat**
- This adds `watchOptions` as entry parallel to `compilerOptions` in tsconfig to specify `watch` specific options. Eg. `watchFile` strategy etc. It would also allow us to set ignores for watches or some such options in future, or allowing delayed project watches in the server and others.
- Currently watchOptions has `watchFile` strategy to watch files, `watchDirectory` strategy to watch individual directory on the OS that don't support native recursive watching and `fallbackPolling` method to poll when fs related watching fails.
- This also adds the delayed update to child watches when the recursive directory watches are invoked. It can be disabled by setting `synchronousWatchDirectory` in `watchOptions`
Fixes #30208, #32033, #34401

### CODE REVIEW

**Comment by sheetalkamat** [comment]
@typescript-bot pack this

**Comment by aranhaagency** [comment]
Nice work!

**Comment by andreialecu** [comment]
Should this help with https://github.com/microsoft/TypeScript/issues/33335 ? If so, are there any aditional steps necessary? I've installed the latest nightly and selected it in VS Code but it doesn't seem to have helped.

```
Info 0    [12:52:10.97] Starting TS Server
Info 1    [12:52:10.98] Version: 3.8.0-dev.20191231
```

File: src/testRunner/unittests/tscWatch/consoleClearing.ts [hunk 1]
```diff
@@ -2,8 +2,8 @@ namespace ts.tscWatch {
     describe("unittests:: tsc-watch:: console clearing", () => {
         const currentDirectoryLog = "Current directory: / CaseSensitiveFileNames: false\n";
         const fileWatcherAddedLog = [
-            "FileWatcher:: Added:: WatchInfo: /f.ts 250 Source file\n",
-            "FileWatcher:: Added:: WatchInfo: /a/lib/lib.d.ts 250 Source file\n"
+            "FileWatcher:: Added:: WatchInfo: /f.ts 250 undefined Source file\n",
```
Comment (DanielRosenwasser) [inline src/testRunner/unittests/tscWatch/consoleClearing.ts]: What does `undefined` refer to here?

Comment (sheetalkamat) [reply src/testRunner/unittests/tscWatch/consoleClearing.ts]: undefined is watchOptions passed to the file watcher

File: src/compiler/commandLineParser.ts [hunk 2]
```diff
@@ -73,6 +73,49 @@ namespace ts {
...
+                dynamicpriority: PollingWatchKind.DynamicPriority,
+            }),
+            category: Diagnostics.Advanced_Options,
+            description: Diagnostics.Specify_strategy_for_creating_a_polling_watch_when_it_fails_to_create_using_file_system_events_Colon_FixedInterval_default_PriorityInterval_DynamicPriority,
+        },
+        {
+            name: "synchronousWatchDirectory",
+            type: "boolean",
+            category: Diagnostics.Advanced_Options,
+            description: Diagnostics.Synchronously_call_callbacks_and_update_the_state_of_directory_watchers_on_platforms_that_don_t_support_recursive_watching_natively,
```
Comment (DanielRosenwasser) [inline src/compiler/commandLineParser.ts]: As a user, is this relevant to me? What do I need to know? It's not clear to me what this option does.

Comment (DanielRosenwasser) [reply src/compiler/commandLineParser.ts]: Also, what's the default?

File: src/compiler/commandLineParser.ts [hunk 3]
```diff
@@ -73,6 +73,49 @@ namespace ts {
...
+            type: createMapFromTemplate({
+                usefsevents: WatchDirectoryKind.UseFsEvents,
+                fixedpollinginterval: WatchDirectoryKind.FixedPollingInterval,
+                dynamicprioritypolling: WatchDirectoryKind.DynamicPriorityPolling,
+            }),
+            category: Diagnostics.Advanced_Options,
+            description: Diagnostics.Specify_strategy_for_watching_directory_on_platforms_that_don_t_support_recursive_watching_natively_Colon_UseFsEvents_default_FixedPollingInterval_DynamicPriorityPolling,
+        },
+        {
+            name: "fallbackPolling",
```
Comment (DanielRosenwasser) [inline src/compiler/commandLineParser.ts]: Is this polling for files, directories, or both?

Comment (sheetalkamat) [reply src/compiler/commandLineParser.ts]: Its for both.. Any fallback when filesystem native fails uses this method

File: src/compiler/commandLineParser.ts (hunk 2 above)
Comment (sheetalkamat) [reply src/compiler/commandLineParser.ts]: default is false and setting this to true would get you behaviour in 3.7.. This is new flag that when not set, on recursive directory watching doesnt update the watcher status right away but instead sets a timeout to update the watches for recursive structure.. (eg. when say removing or creating node_modules folder, while npm install is going, the recursive watchers aren't updated right away but on timeout so we aren't eg adding watches on directories that will anyways be deleted or recreated..)

Comment (DanielRosenwasser) [reply src/compiler/commandLineParser.ts]: Who would you say would use this feature? Here's how I'm describing it.

> Disable deferred watching on directories. Deferred watching is useful when lots of file changes might occur at once (e.g. a change in `node_modules` from `npm install`), but you might want to disable it with this flag for some atypical setups.

Comment (sheetalkamat) [reply src/compiler/commandLineParser.ts]: This is kind of backup flag that can disable this behavior if changes aren't reflected correctly/sooner.
